// pixi-projection/lib/index.js
'use strict';

require('./base/index.js');
var utils = require('./utils.js');
require('./proj2d/index.js');
var Point3d = require('./proj3d/Point3d.js');
var Euler = require('./proj3d/Euler.js');
var ObservableEuler = require('./proj3d/ObservableEuler.js');
var Matrix3d = require('./proj3d/Matrix3d.js');
var Projection3d = require('./proj3d/Projection3d.js');
var Container3d = require('./proj3d/Container3d.js');
var Camera3d = require('./proj3d/Camera3d.js');
require('./proj3d/sprites/convert.js');
var Sprite3d = require('./proj3d/sprites/Sprite3d.js');
var Text3d = require('./proj3d/sprites/Text3d.js');
var Mesh3d2d = require('./proj3d/mesh/Mesh3d2d.js');
require('./curve/index.js');
var Spine2d = require('./spine/Spine2d.js');
var Spine3d = require('./spine/Spine3d.js');
var AbstractProjection = require('./base/AbstractProjection.js');
var LinearProjection = require('./base/LinearProjection.js');
var Sprite2dRenderer = require('./base/webgl/Sprite2dRenderer.js');
var UniformBatchRenderer = require('./base/webgl/UniformBatchRenderer.js');
var Matrix2d = require('./proj2d/Matrix2d.js');
var Projection2d = require('./proj2d/Projection2d.js');
var Container2d = require('./proj2d/Container2d.js');
var Sprite2d = require('./proj2d/sprites/Sprite2d.js');
var Text2d = require('./proj2d/sprites/Text2d.js');
var TilingSprite2d = require('./proj2d/tiling/TilingSprite2d.js');
var TilingSprite2dRenderer = require('./proj2d/tiling/TilingSprite2dRenderer.js');
var MaskHacker = require('./proj2d/z_masks/MaskHacker.js');
var SpriteMaskFilter = require('./proj2d/z_masks/SpriteMaskFilter.js');
var Mesh2d = require('./proj2d/mesh/Mesh2d.js');
var BaseSurface = require('./curve/BaseSurface.js');
var BilinearSurface = require('./curve/BilinearSurface.js');
var ProjectionSurface = require('./curve/ProjectionSurface.js');
var SpriteBilinearRenderer = require('./curve/SpriteBilinearRenderer.js');
var Sprite2s = require('./curve/sprites/Sprite2s.js');
var Text2s = require('./curve/sprites/Text2s.js');



exports.getIntersectionFactor = utils.getIntersectionFactor;
exports.getPositionFromQuad = utils.getPositionFromQuad;
exports.ObservablePoint3d = Point3d.ObservablePoint3d;
exports.Point3d = Point3d.Point3d;
exports.Euler = Euler.Euler;
exports.ObservableEuler = ObservableEuler.ObservableEuler;
exports.Matrix3d = Matrix3d.Matrix3d;
exports.Projection3d = Projection3d.Projection3d;
exports.Container3d = Container3d.Container3d;
exports.container3dGetDepth = Container3d.container3dGetDepth;
exports.container3dIsFrontFace = Container3d.container3dIsFrontFace;
exports.container3dToLocal = Container3d.container3dToLocal;
exports.container3dWorldTransform = Container3d.container3dWorldTransform;
exports.Camera3d = Camera3d.Camera3d;
exports.Sprite3d = Sprite3d.Sprite3d;
exports.Text3d = Text3d.Text3d;
exports.Mesh3d2d = Mesh3d2d.Mesh3d2d;
exports.SimpleMesh3d2d = Mesh3d2d.SimpleMesh3d2d;
exports.applySpine2dMixin = Spine2d.applySpine2dMixin;
exports.applySpine3dMixin = Spine3d.applySpine3dMixin;
exports.AbstractProjection = AbstractProjection.AbstractProjection;
exports.TRANSFORM_STEP = AbstractProjection.TRANSFORM_STEP;
exports.AFFINE = LinearProjection.AFFINE;
exports.LinearProjection = LinearProjection.LinearProjection;
exports.transformHack = LinearProjection.transformHack;
exports.Batch2dRenderer = Sprite2dRenderer.Batch2dRenderer;
exports.Batch3dGeometry = Sprite2dRenderer.Batch3dGeometry;
exports.UniformBatchRenderer = UniformBatchRenderer.UniformBatchRenderer;
exports.Matrix2d = Matrix2d.Matrix2d;
exports.Projection2d = Projection2d.Projection2d;
exports.Container2d = Container2d.Container2d;
exports.container2dToLocal = Container2d.container2dToLocal;
exports.container2dWorldTransform = Container2d.container2dWorldTransform;
exports.Sprite2d = Sprite2d.Sprite2d;
exports.Text2d = Text2d.Text2d;
exports.TilingSprite2d = TilingSprite2d.TilingSprite2d;
exports.TilingSprite2dRenderer = TilingSprite2dRenderer.TilingSprite2dRenderer;
exports.patchSpriteMask = MaskHacker.patchSpriteMask;
exports.SpriteMaskFilter2d = SpriteMaskFilter.SpriteMaskFilter2d;
exports.Mesh2d = Mesh2d.Mesh2d;
exports.SimpleMesh2d = Mesh2d.SimpleMesh2d;
exports.Surface = BaseSurface.Surface;
exports.BilinearSurface = BilinearSurface.BilinearSurface;
exports.ProjectionSurface = ProjectionSurface.ProjectionSurface;
exports.BatchBilinearGeometry = SpriteBilinearRenderer.BatchBilinearGeometry;
exports.BatchBilinearRenderer = SpriteBilinearRenderer.BatchBilinearRenderer;
exports.Sprite2s = Sprite2s.Sprite2s;
exports.Text2s = Text2s.Text2s;
