// pixi-projection/lib/index.mjs
import './base/index.mjs';
export { getIntersectionFactor, getPositionFromQuad } from './utils.mjs';
import './proj2d/index.mjs';
export { ObservablePoint3d, Point3d } from './proj3d/Point3d.mjs';
export { Euler } from './proj3d/Euler.mjs';
export { ObservableEuler } from './proj3d/ObservableEuler.mjs';
export { Matrix3d } from './proj3d/Matrix3d.mjs';
export { Projection3d } from './proj3d/Projection3d.mjs';
export { Container3d, container3dGetDepth, container3dIsFrontFace, container3dToLocal, container3dWorldTransform } from './proj3d/Container3d.mjs';
export { Camera3d } from './proj3d/Camera3d.mjs';
import './proj3d/sprites/convert.mjs';
export { Sprite3d } from './proj3d/sprites/Sprite3d.mjs';
export { Text3d } from './proj3d/sprites/Text3d.mjs';
export { Mesh3d2d, SimpleMesh3d2d } from './proj3d/mesh/Mesh3d2d.mjs';
import './curve/index.mjs';
export { applySpine2dMixin } from './spine/Spine2d.mjs';
export { applySpine3dMixin } from './spine/Spine3d.mjs';
export { AbstractProjection, TRANSFORM_STEP } from './base/AbstractProjection.mjs';
export { AFFINE, LinearProjection, transformHack } from './base/LinearProjection.mjs';
export { Batch2dRenderer, Batch3dGeometry } from './base/webgl/Sprite2dRenderer.mjs';
export { UniformBatchRenderer } from './base/webgl/UniformBatchRenderer.mjs';
export { Matrix2d } from './proj2d/Matrix2d.mjs';
export { Projection2d } from './proj2d/Projection2d.mjs';
export { Container2d, container2dToLocal, container2dWorldTransform } from './proj2d/Container2d.mjs';
export { Sprite2d } from './proj2d/sprites/Sprite2d.mjs';
export { Text2d } from './proj2d/sprites/Text2d.mjs';
export { TilingSprite2d } from './proj2d/tiling/TilingSprite2d.mjs';
export { TilingSprite2dRenderer } from './proj2d/tiling/TilingSprite2dRenderer.mjs';
export { patchSpriteMask } from './proj2d/z_masks/MaskHacker.mjs';
export { SpriteMaskFilter2d } from './proj2d/z_masks/SpriteMaskFilter.mjs';
export { Mesh2d, SimpleMesh2d } from './proj2d/mesh/Mesh2d.mjs';
export { Surface } from './curve/BaseSurface.mjs';
export { BilinearSurface } from './curve/BilinearSurface.mjs';
export { ProjectionSurface } from './curve/ProjectionSurface.mjs';
export { BatchBilinearGeometry, BatchBilinearRenderer } from './curve/SpriteBilinearRenderer.mjs';
export { Sprite2s } from './curve/sprites/Sprite2s.mjs';
export { Text2s } from './curve/sprites/Text2s.mjs';
