// pixi-projection/lib/proj2d/Container2d.mjs
import { Projection2d } from './Projection2d.mjs';
import { Container } from '@pixi/display';
import '../base/index.mjs';
import { TRANSFORM_STEP } from '../base/AbstractProjection.mjs';

function container2dWorldTransform() {
  return this.proj.affine ? this.transform.worldTransform : this.proj.world;
}
class Container2d extends Container {
  constructor() {
    super();
    this.proj = new Projection2d(this.transform);
  }
  toLocal(position, from, point, skipUpdate, step = TRANSFORM_STEP.ALL) {
    if (from) {
      position = from.toGlobal(position, point, skipUpdate);
    }
    if (!skipUpdate) {
      this._recursivePostUpdateTransform();
    }
    if (step >= TRANSFORM_STEP.PROJ) {
      if (!skipUpdate) {
        this.displayObjectUpdateTransform();
      }
      if (this.proj.affine) {
        return this.transform.worldTransform.applyInverse(position, point);
      }
      return this.proj.world.applyInverse(position, point);
    }
    if (this.parent) {
      point = this.parent.worldTransform.applyInverse(position, point);
    } else {
      point.x = position.x;
      point.y = position.y;
    }
    if (step === TRANSFORM_STEP.NONE) {
      return point;
    }
    return this.transform.localTransform.applyInverse(point, point);
  }
  get worldTransform() {
    return this.proj.affine ? this.transform.worldTransform : this.proj.world;
  }
}
const container2dToLocal = Container2d.prototype.toLocal;

export { Container2d, container2dToLocal, container2dWorldTransform };
