// pixi-projection/lib/proj2d/index.js
'use strict';

var core = require('@pixi/core');
var TilingSprite2dRenderer = require('./tiling/TilingSprite2dRenderer.js');
var Matrix2d = require('./Matrix2d.js');
var Projection2d = require('./Projection2d.js');
var Container2d = require('./Container2d.js');
require('./sprites/convert.js');
require('@pixi/sprite');
require('../base/index.js');
var Text2d = require('./sprites/Text2d.js');
var TilingSprite2d = require('./tiling/TilingSprite2d.js');
var MaskHacker = require('./z_masks/MaskHacker.js');
var Mesh2d = require('./mesh/Mesh2d.js');

core.extensions.add(TilingSprite2dRenderer.TilingSprite2dRenderer);

exports.TilingSprite2dRenderer = TilingSprite2dRenderer.TilingSprite2dRenderer;
exports.Matrix2d = Matrix2d.Matrix2d;
exports.Projection2d = Projection2d.Projection2d;
exports.Container2d = Container2d.Container2d;
exports.container2dToLocal = Container2d.container2dToLocal;
exports.container2dWorldTransform = Container2d.container2dWorldTransform;
exports.Text2d = Text2d.Text2d;
exports.TilingSprite2d = TilingSprite2d.TilingSprite2d;
exports.patchSpriteMask = MaskHacker.patchSpriteMask;
exports.Mesh2d = Mesh2d.Mesh2d;
exports.SimpleMesh2d = Mesh2d.SimpleMesh2d;
